### ISSUE TITLE
Sometimes when you release some item it goes to the wrong place.

### ISSUE THREAD

**Issue by marcglasberg** [bnxm/implicitly_animated_reorderable_list#23]
Suppose you have a vertical, top-down list with, say, 10 items:

• Go grab item 6 and slowly move it up.
• Item 5 will go down and open some space between items 4 and 5.
• Continue moving item 6 up.
• Now item 4 will go down and open some space between items 3 and 4.
• Continue moving item 6 up, but stop when the top of item 6 is a little bit above the bottom of item 3.
• Then release item 6.
• At this point there was some space between items 3 and 4, and that's where item 6 should land.
• However, as soon as you release it, it will land between items 2 and 3.

It seems to me that, for some reason, the calculation on where the item should land is different from the calculation during the drag itself. So, when you release it, it thinks it should go to a different place.

This is kind of a serious bug, because it is very disconcerting that the item should move to a place different than the one you release it. This makes the list just feel "wrong" sometimes, and this problem seems to happen in both directions, only when the edge of the moving item is a little bit over the edge of the next item.

**Comment by marcglasberg** [comment]
Or, I'm not sure this is a real effect. Maybe it's just a usability problem.

In the above example, if you want to move item 6 to land between items 3 and 4, users naturally try to move the item to the **space between items 3 and 4**. When using the real device they will aim there and then execute the movement very quickly. However, at this location, the item is very close to the location where it could move item 3 down.

The implicitly_animated_reorderable_list in fact expects the user to move the item not to the space between items 3 and 4, but, instead, to to top of the item you want to take its place. So, in this example, if you want to move item 6 to land between items 3 and 4, it works OK if the user moves the item to the top of item 4.

Unfortunately,  as I said, there is not where users aim, in practice, during our usability tests. So it just "feels" wrong.

When using the emulator this doesn't happen, because then you have a big screen, and you move it more slowly (instead of aiming and quickly moving to the desired place) and then you have time to see the items moving and the space opening up.

**Comment by marcglasberg** [comment]
Let's say you have two mode of movement: **1) On top**, and **2) In the middle**.

1) **On top**: This is the situation you have today. The user must aim to release the item on top of the item it wants to displace to the direction opposite to the movement. So, for example, moving item 6 to the top of item 3 will move item 3 **down**, but moving item 1 in top of item 3 will move item 3 **up**.

1) **In the middle**: This would be much better for usability. The user must aim to release the item on the middle of the items. So, for example, moving any item to the space between items 3 and 4 will always result in that item landing between them.

**Comment by marcglasberg** [comment]
A question: Is this movement logic your own, or was it copied from https://api.flutter.dev/flutter/material/ReorderableListView-class.html ?

**Comment by bnxm** [comment]
No, this is my own logic and I agree with the points you brought up however I'm currently relatively busy so I'm not sure when I find the time to work on it.

**Comment by marcglasberg** [comment]
@bnxm I understand and can simpatize with the problem of finding the time to maintain open source projects. Well, I hope you find the time soon. Thanks.

**Comment by killian2k** [comment]
Hello, 
I have the exact same issue. 
Do you know when you should be able to fix this issue?

**Comment by wonpyohong** [comment]
I have same issue

**Comment by leopries** [comment]
@bnxm Will this be fixed any time soon? It makes the reorderable List unusable

**Comment by DomingoMG** [comment]
It seems to give a problem, I am going to implement my own list.

**Comment by wonpyohong** [comment]
I debugged with log and I found something.
```
onPointerMove: (event) {
        final pointer = event.localPosition;
        final delta = _isVertical ? event.delta.dy : event.delta.dx;

        if (_inDrag && _inReorder) {
          _onDragUpdated(pointer, delta.isNegative);
        }
      },
```

'delta.isNegative' is sensitively changing value and start point of bug.
Even though we 200px up, delta can be randomly negative when finger release.

in onDragEnded() if delta is negative, ListView recognize that it is downward.

For example, 4 items positions are below
0: 0~72, center: 36
1: 72~144, center: 108
2: 144~216, center: 180

I will drag item 2 and drag up to position 109 (top position).
by now there is no animation because 109 > 108 (still below item 2's center)
When we release finger, our expected behavior is no reorder.
However, unluckly, if pointer go down slightly to 110 on releasing finger, it will be reordered.

why reorder?
position is 110 and moving downward -> closest item is 1 -> 

```
void onDragEnded() {
  ..
  // _next means closest item. here item 1.
  final delta = _next != dragItem ? _next.start - _dragStart : -_pointerDelta;     // _next.start - _dragStart = item 1's start - 110 = -38
  _dispatchMove(dragKey, delta, ..)
  ..
```

so dragging item moves -38 (up)

**Comment by marcglasberg** [comment]
@wonpyohong Can you suggest a fix?

**Comment by wonpyohong** [comment]
@marcglasberg 
I am looking into it comparing with android RecyclerView. 
I guess that it can be hard to fix with a small change.
I don't know yet it can be fixed fundamentally with _closestList basis

**Comment by wonpyohong** [comment]
@marcglasberg 
I fixed and send PR. But this package is not managed well I guess. 
You can copy & paste and apply changes to your own code if you want to apply it ASAP.
Let me know if there are bugs

**Comment by bnxm** [comment]
@wonpyohong Thanks for trying to fix the issue yourself. It looks pretty good! I just added one minor addition though, that I think will fully fix this issue.

**Comment by bnxm** [comment]
@marcglasberg
@wonpyohong

I made a pre-release (0.3.0-dev+1) with those changes included. I will release it to the production version until next week, if no other issues come up in the meantime.

**Comment by marcglasberg** [comment]
It's really perfect now!
